cpp/bench/ann/src/common/dataset.h

```diff
@@ -0,0 +1,381 @@
+void BinFile<T>::check_suffix_()
+{
+  auto pos = file_.rfind('.');
+  if (pos == std::string::npos) {
+    throw std::runtime_error("name of BinFile doesn't have a suffix: " + file_);
+  }
+  std::string suffix = file_.substr(pos + 1);
+
+  if constexpr (std::is_same_v<T, float>) {
+    if (suffix != "fbin") {
+      throw std::runtime_error("BinFile<float> should has .fbin suffix: " + file_);
+    }
+  } else if constexpr (std::is_same_v<T, half>) {
+    if (suffix != "f16bin") {
+      throw std::runtime_error("BinFile<half> should has .f16bin suffix: " + file_);
+    }
+  } else if constexpr (std::is_same_v<T, int>) {
+    if (suffix != "ibin") {
+      throw std::runtime_error("BinFile<int> should has .ibin suffix: " + file_);
+    }
+  } else if constexpr (std::is_same_v<T, uint8_t>) {
+    if (suffix != "u8bin") {
+      throw std::runtime_error("BinFile<uint8_t> should has .u8bin suffix: " + file_);
+    }
+  } else if constexpr (std::is_same_v<T, int8_t>) {
+    if (suffix != "i8bin") {
+      throw std::runtime_error("BinFile<int8_t> should has .i8bin suffix: " + file_);
+    }
+  } else {
+    throw std::runtime_error(
+      "T of BinFile<T> should be one of float, half, int, uint8_t, or int8_t");
+  }
+}
+
+template <typename T>
+class Dataset {
+ public:
+  Dataset(const std::string& name) : name_(name) {}
+  Dataset(const std::string& name, const std::string& distance) : name_(name), distance_(distance)
+  {
+  }
+  Dataset(const Dataset&) = delete;
+  Dataset& operator=(const Dataset&) = delete;
+  virtual ~Dataset();
+
+  std::string name() const { return name_; }
+  std::string distance() const { return distance_; }
+  int dim() const { return dim_; }
+  size_t base_set_size() const { return base_set_size_; }
+  size_t query_set_size() const { return query_set_size_; }
+
+  // load data lazily, so don't pay the overhead of reading unneeded set
+  // e.g. don't load base set when searching
+  const T* base_set() const
+  {
+    if (!base_set_) { load_base_set_(); }
+    return base_set_;
+  }
+
+  const T* query_set() const
+  {
+    if (!query_set_) { load_query_set_(); }
+    return query_set_;
+  }
+
+  const T* base_set_on_gpu() const;
+  const T* query_set_on_gpu() const;
+  const T* mapped_base_set() const;
+
+ protected:
+  virtual void load_base_set_() const  = 0;
+  virtual void load_query_set_() const = 0;
+  virtual void map_base_set_() const   = 0;
+
+  std::string name_;
+  std::string distance_;
+  int dim_;
+  size_t base_set_size_;
+  size_t query_set_size_;
+
+  mutable T* base_set_        = nullptr;
+  mutable T* query_set_       = nullptr;
+  mutable T* d_base_set_      = nullptr;
+  mutable T* d_query_set_     = nullptr;
+  mutable T* mapped_base_set_ = nullptr;
+};
+
+template <typename T>
+Dataset<T>::~Dataset()
+{
+  delete[] base_set_;
+  delete[] query_set_;
+  if (d_base_set_) { RAFT_CUDA_TRY_NO_THROW(cudaFree(d_base_set_)); }
+  if (d_query_set_) { RAFT_CUDA_TRY_NO_THROW(cudaFree(d_query_set_)); }
```

Comment:
Rather than managing our own memory - can we use things like rmm::device_uvector etc here (and likewise some kind of smart pointer for the host memory)?

Reply:
Created https://github.com/rapidsai/raft/issues/1367 to do this. I'm just a little wary of doing it here because I don't want to break things. I'd like to do it in the follow-ups